python/tvm/relay/qnn/op/legalizations.py

```diff
@@ -35,32 +50,51 @@ def legalize_qnn_conv2d(attrs, inputs, types):
         The args of the Relay expr to be legalized
     types : list of types
         List of input and output types
+    op : Relay op
+        The original Relay op.
 
     Returns
     -------
     result : tvm.relay.Expr
         The legalized expr
+
     """
-    return qnn_conv2d_legalize(attrs, inputs, types)
+    # Collect the dtypes.
+    data_dtype = types[0].dtype
+    kernel_dtype = types[1].dtype
 
-# Generic QNN Conv2D legalization function.
-@tvm.target.generic_func
-def qnn_conv2d_legalize(attrs, inputs, types):
-    """Default legalization is None."""
-    return None
+    # Collect the input exprs.
+    data, kernel = inputs
+
+    # Return if the data types are same.
+    if data_dtype == kernel_dtype:
+        return None
+
+    # Make input dtype same as kernel dtype.
+    data = shift_zero_point(data, kernel_dtype)
+    input_zp = attrs['input_zero_point']
+    if data_dtype == 'int8':
+        input_zp = input_zp + 128
+    elif data_dtype == 'uint8':
+        input_zp = input_zp - 128
+    else:
+        return None
```

Comment:
i think we should throw an error here. this is quantization specfic legalization and no other data type should be passed.

Reply:
The datatype could be int32. I can add a check in the start to look for legalizations only when the data type is 8-bit, and put an assertion here.